distributed/diagnostics/plugin.py

```diff
@@ -376,204 +376,162 @@ class PackageInstall(SchedulerPlugin, abc.ABC):
-            else:
-                logger.info(
-                    "The following packages have already been installed: %s",
-                    self.installer.packages,
-                )
+            self._install_fn()
 
-            if self.restart and worker.nanny and not await self._is_restarted(worker):
-                logger.info("Restarting worker to refresh interpreter.")
-                await self._set_restarted(worker)
-                worker.loop.add_callback(
-                    worker.close_gracefully, restart=True, reason=f"{self.name}-setup"
-                )
 
-    async def _is_installed(self, worker):
-        return await worker.client.get_metadata(
-            self._compose_installed_key(), default=False
-        )
+class _InstallWorkerPlugin(WorkerPlugin):
+    """Worker plugin to install software on the worker
 
-    async def _set_installed(self, worker):
-        await worker.client.set_metadata(
-            self._compose_installed_key(),
-            True,
-        )
+    This accepts an installer which will installs software on all workers.
 
-    def _compose_installed_key(self):
-        return [
-            self.name,
-            "installed",
-            socket.gethostname(),
-        ]
-
-    async def _is_restarted(self, worker):
-        return await worker.client.get_metadata(
-            self._compose_restarted_key(worker),
-            default=False,
-        )
+    .. note::
 
-    async def _set_restarted(self, worker):
-        await worker.client.set_metadata(
-            self._compose_restarted_key(worker),
-            True,
-        )
+       This will increase the time it takes to start up
+       the cluster. If possible, we recommend including the
+       software in the cluster environment or image. This is
+       primarily intended for experimentation and debugging.
 
-    def _compose_restarted_key(self, worker):
-        return [self.name, "restarted", worker.nanny]
+    Parameters
+    ----------
+    install_fn
+        Callable that installs software
+    name:
+        Name of the plugin
 
+    See Also
+    --------
+    InstallPlugin
+    """
 
-class _PackageInstaller(abc.ABC):
-    INSTALLER: ClassVar[str]
-    packages: list[str]
+    _install_fn: Callable[[], None]
+    name: str
 
-    def __init__(self, packages: list[str]):
-        self.packages = packages
+    def __init__(self, install_fn: Callable[[], None], name: str):
+        self._install_fn = install_fn
+        self.name = name
 
-    @abc.abstractmethod
-    def install(self) -> None:
-        """Install the requested packages"""
+    async def setup(self, worker):
+        from distributed.semaphore import Semaphore
 
+        async with (
+            await Semaphore(
+                max_leases=1,
+                name=socket.gethostname(),
+                register=True,
+                scheduler_rpc=worker.scheduler,
+                loop=worker.loop,
+            )
```

Comment:
unrelated to the change in this PR but reviewing this again I wonder if a lock **file** wouldn't be much simpler.
We're typically using `locket` for this, e.g. https://github.com/dask/distributed/blob/e98dcb19779218cace5de69310377e11933e09a5/distributed/diskutils.py#L161

Reply:
Good point, I hadn't thought of this. This would also solve the issue that using a Semaphore from the scheduler isn't straightforward (if not even impossible).

Reply:
I'd leave this to a follow-up PR though.